File: frontend/financeiro/lib/app/data/model/talonario_cheque_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:financeiro/app/data/model/model_imports.dart';
import 'package:financeiro/app/data/domain/domain_imports.dart';

class TalonarioChequeModel {
	int? id;
	int? idBancoContaCaixa;
	String? talao;
	int? numero;
	String? statusTalao;
	List<ChequeModel>? chequeModelList;
	BancoContaCaixaModel? bancoContaCaixaModel;

	TalonarioChequeModel({
		this.id,
		this.idBancoContaCaixa,
		this.talao,
		this.numero,
		this.statusTalao,
		this.chequeModelList,
		this.bancoContaCaixaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'talao',
		'numero',
		'status_talao',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Talao',
		'Numero',
		'Status Talao',
	];

	TalonarioChequeModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idBancoContaCaixa = jsonData['idBancoContaCaixa'];
		talao = jsonData['talao'];
		numero = jsonData['numero'];
		statusTalao = TalonarioChequeDomain.getStatusTalao(jsonData['statusTalao']);
		chequeModelList = (jsonData['chequeModelList'] as Iterable?)?.map((m) => ChequeModel.fromJson(m)).toList() ?? [];
		bancoContaCaixaModel = jsonData['bancoContaCaixaModel'] == null ? BancoContaCaixaModel() : BancoContaCaixaModel.fromJson(jsonData['bancoContaCaixaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idBancoContaCaixa'] = idBancoContaCaixa != 0 ? idBancoContaCaixa : null;
		jsonData['talao'] = talao;
		jsonData['numero'] = numero;
		jsonData['statusTalao'] = TalonarioChequeDomain.setStatusTalao(statusTalao);
		
		var chequeModelLocalList = []; 
		for (ChequeModel object in chequeModelList ?? []) { 
			chequeModelLocalList.add(object.toJson); 
		}
		jsonData['chequeModelList'] = chequeModelLocalList;
		jsonData['bancoContaCaixaModel'] = bancoContaCaixaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idBancoContaCaixa = plutoRow.cells['idBancoContaCaixa']?.value;
		talao = plutoRow.cells['talao']?.value;
		numero = plutoRow.cells['numero']?.value;
		statusTalao = plutoRow.cells['statusTalao']?.value != '' ? plutoRow.cells['statusTalao']?.value : 'Normal';
		chequeModelList = [];
		bancoContaCaixaModel = BancoContaCaixaModel();
		bancoContaCaixaModel?.nome = plutoRow.cells['bancoContaCaixaModel']?.value;
	}	

	TalonarioChequeModel clone() {
		return TalonarioChequeModel(
			id: id,
			idBancoContaCaixa: idBancoContaCaixa,
			talao: talao,
			numero: numero,
			statusTalao: statusTalao,
			chequeModelList: chequeModelListClone(chequeModelList!),
		);			
	}

	chequeModelListClone(List<ChequeModel> chequeModelList) { 
		List<ChequeModel> resultList = [];
		for (var chequeModel in chequeModelList) {
			resultList.add(
				ChequeModel(
					id: chequeModel.id,
					idTalonarioCheque: chequeModel.idTalonarioCheque,
					numero: chequeModel.numero,
					statusCheque: chequeModel.statusCheque,
					dataStatus: chequeModel.dataStatus,
				)
			);
		}
		return resultList;
	}

	
}

File: backend/financeiro/src/model/talonario_cheque_model.py
from src import db
from src.model.cheque_model import ChequeModel
from src.model.banco_conta_caixa_model import BancoContaCaixaModel


class TalonarioChequeModel(db.Model):
    __tablename__ = 'talonario_cheque'

    id = db.Column(db.Integer, primary_key=True)
    talao = db.Column(db.String(10))
    numero = db.Column(db.Integer)
    status_talao = db.Column(db.String(1))
    id_banco_conta_caixa = db.Column(db.Integer, db.ForeignKey('banco_conta_caixa.id'))

    cheque_model_list = db.relationship('ChequeModel', lazy='dynamic')
    banco_conta_caixa_model = db.relationship('BancoContaCaixaModel', foreign_keys=[id_banco_conta_caixa])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_banco_conta_caixa = data.get('idBancoContaCaixa')
        self.talao = data.get('talao')
        self.numero = data.get('numero')
        self.status_talao = data.get('statusTalao')

    def serialize(self):
        return {
            'id': self.id,
            'idBancoContaCaixa': self.id_banco_conta_caixa,
            'talao': self.talao,
            'numero': self.numero,
            'statusTalao': self.status_talao,
            'chequeModelList': [cheque_model.serialize() for cheque_model in self.cheque_model_list],
            'bancoContaCaixaModel': self.banco_conta_caixa_model.serialize() if self.banco_conta_caixa_model else None,
        }